[upgrade guide] [i18nIgnore] Add upgrade to v2 guide + moving pieces (#2330)
withastro/docs: src/pages/en/guides/upgrade-to/v2.mdx

##### What should I do?

Remove `extendDefaultPlugins: true` in your configuration. This is now Astro's default behavior in 2.0, and you can delete this line without any replacement.

```js del={6}
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  markdown: {
    extendDefaultPlugins: true,
  }
});
```

#### Added: `gfm` and `smartypants`

Astro 2.0 replaces `markdown.extendDefaultPlugins` with separate Boolean options to control each of Astro's built-in default Markdown plugins individually. These are enabled by default, and can be set to `false` to remove each plugin independently.

##### What should I do?

Replace `extendDefaultPlugins: false` and add the flags to disable each plugin individually instead.

- `markdown.gfm: false` disables GitHub-Flavored Markdown
- `markdown.smartypants: false` disables SmartyPants

```js del={6} ins={7-8}
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  markdown: {
    extendDefaultPlugins: false,
    smartypants: false,
    gfm: false,
  }
});
```

### Changed: MDX plugin configuration

#### Replaced: `extendPlugins` -> `extendMarkdownConfig`

> In 1.0, `.mdx.extendPlugins` managed how your MDX files should inherit your Markdown plugin configuration. This setting was used to indicate whether all Markdown plugins (`markdown`), or only Astro's default plugins (`defaults`) should be applied to MDX.
> 
> Astro 2.0 replaces `mdx.extendPlugins` with three new, independently-configurable options:
> 
> - **[`mdx.extendMarkdownConfig`](/en/guides/integrations-guide/mdx/#extendmarkdownconfig)** (`true` by default) to inherit all of your Markdown configuration or none of it.
> - **`mdx.gfm`** (`true` by default) to enable or disable GitHub-Flavored Markdown in MDX.
> - **`mdx.smartypants`** (`true` by default) to enable or disable SmartyPants in MDX.
> 
>  ##### What should I do?
> 
> Delete `extendPlugins: 'markdown'` in your configuration. This is now default behavior. 
> 
> ```js del={6}
> // astro.config.mjs
> import { defineConfig } from 'astro/config';
> 
> export default defineConfig({
>   mdx: {
>     extendPlugins: markdown,
>   }
> });
> ```
> 
> Replace `extendPlugins: 'defaults'` with `extendMarkdownConfig: false` and add the [separate options for GitHub-Flavored Markdown and SmartyPants](/en/guides/integrations-guide/mdx/#options-inherited-from-markdown-config) to enable these default plugins individually in MDX.
> 
> ```js del={6} ins={7-9}
> // astro.config.mjs
> import { defineConfig } from 'astro/config';
> 
> export default defineConfig({
>   mdx: {
>     extendPlugins: defaults,
>     extendMarkdownConfig: false,
>     smartypants: true,
>     gfm: true,
>   }
> });
> ```

Reviewer: These options are passed to the MDX integration not to an `mdx` config key, so reworked this to reflect that:

````suggestion
In 1.0, the MDX integration’s `extendPlugins` option managed how your MDX files should inherit your Markdown configuration.

Astro 2.0 replaces `extendPlugins` with three new, independently-configurable options:

- **[`extendMarkdownConfig`](/en/guides/integrations-guide/mdx/#extendmarkdownconfig)** (`true` by default) to inherit all of your Markdown configuration or none of it.
- **`gfm`** (`true` by default) to enable or disable GitHub-Flavored Markdown in MDX.
- **`smartypants`** (`true` by default) to enable or disable SmartyPants in MDX.

##### What should I do?

Delete `extendPlugins: 'markdown'` in your configuration. This is now default behavior. 

```js del={8}
// astro.config.mjs
import { defineConfig } from 'astro/config';
import mdx from '@astrojs/mdx';

export default defineConfig({
  integrations: [
    mdx({
      extendPlugins: 'markdown',
    }),
  ],
});
```

Replace `extendPlugins: 'defaults'` with `extendMarkdownConfig: false` and add the [separate options for GitHub-Flavored Markdown and SmartyPants](/en/guides/integrations-guide/mdx/#options-inherited-from-markdown-config) to enable these default plugins individually in MDX.

```js del={8} ins={9-11}
// astro.config.mjs
import { defineConfig } from 'astro/config';
import mdx from '@astrojs/mdx';

export default defineConfig({
  integrations: [
    mdx({
      extendPlugins: 'defaults',
      extendMarkdownConfig: false,
      smartypants: true,
      gfm: true,
    }),
  ],
});
```
````

#### Added: More MDX config options to match Markdown

> Setting `extendMarkdownConfig: false` now allows you to individually set [every available Markdown configuration option](/en/reference/configuration-reference/#markdown-options) (except `drafts`) separately in your MDX integration configuration.

Reviewer: This section is confusing. The text says you can set `extendMarkdownConfig: false` but the code sample doesn’t show where and I don’t really understand the combo of inherited/ignored/applied/overridden in the comments.

```js
// astro.config.mjs
import { defineConfig } from 'astro/config';
import mdx from '@astrojs/mdx';

export default defineConfig({
  markdown: {
    syntaxHighlight: 'prism',
    remarkPlugins: [remarkPlugin1],
    gfm: true,
  },
  integrations: [
    mdx({
      // `syntaxHighlight` inherited from Markdown

      // Markdown `remarkPlugins` ignored,
      // only `remarkPlugin2` applied.
      remarkPlugins: [remarkPlugin2],
      // `gfm` overridden to `false`
      gfm: false,
    })
  ]
});
```

##### What should I do?

Revisit your Markdown and MDX configuration and compare your existing config with the new options available.

> :::tip[remember]
> - Astro's two default Markdown plugins are now individually configurable, and are not removed when you add other plugins. No more extending defaults!
> - The option to extend your Markdown plugin config to MDX has moved from a selection (just defaults or all plugins) to toggle (all or none).
> - Every Markdown config option (except drafts) can also be configured independently in MDX!
> :::

Reviewer: Tried reworking this but… I can understand the first two too well and aren’t we just repeating what’s above? What do you think about removing this?

```suggestion
```

### Added: Plugin access to frontmatter

In 1.0, remark and rehype plugins did not have access to user frontmatter. Astro merged plugin frontmatter with your file's frontmatter, without passing the file frontmatter to your plugins.

Astro 2.0 gives remark and rehype plugins access to user frontmatter via frontmatter injection. This allows plugin authors to modify a user's existing frontmatter, or compute new properties based on other properties.

#### What should I do?

Check any remark and rehype plugins you have written to see whether their behavior has changed. Note that `data.astro.frontmatter` is now the _complete_ Markdown or MDX document's frontmatter, rather than an empty object.

### Reserved: `src/content/`

> In 1.0, the [Collections API was hidden behind an experimental flag](#astro-20-experimental-flags-removed). This API reserves `src/content/` as a special folder.
> 
> Astro 2.0 now includes the Collections API in the core codebase for organizing your Markdown and MDX files into [content collections](/en/guides/content-collections/).

Reviewer: Do you think the history is important here given how relatively brief the experimental period was? Not sure it’s particularly relevant to users.

```suggestion
Astro 2.0 now includes the Collections API for organizing your Markdown and MDX files into [content collections](/en/guides/content-collections/). This API reserves `src/content/` as a special folder.
```

Author: Right. I was trying to think of a "previously" statement to match the pattern of the others. But, no need to force it!

#### What should I do?

Rename an existing `src/content/` folder to avoid conflicts. This folder, if it exists, can now only be used for [content collections](/en/guides/content-collections/).

## Astro 2.0 Experimental Flags Removed

> Astro v2.0 brings some new features from experimental flags into the core codebase.

Reviewer: Nit: feel like the language isn’t quite right here but not particularly happy with the double “feature” here either.

```suggestion
Astro v2.0 makes some new features that were previously behind experimental flags part of its core feature set.
```

Author: Deleted this line, to see whether we think it's good enough without!

#### What should I do?

Remove the following experimental flags from `astro.config.mjs`:

> ```js del={5-7}
> // astro.config.mjs
> import { defineConfig } from 'astro/config';
> 
> export default defineConfig({
>   experimental: { contentCollections: true },
>   experimental: { prerender: true },
>   experimental: { error-overlay: true },
> })
> ```

Reviewer: Given you wouldn’t have three `experimental` entries in your config, maybe it’s better to show this this way? (Also corrected the error overlay key.)

````suggestion
```js del={5-8}
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  experimental: {
    contentCollections: true,
    prerender: true,
    errorOverlay: true,
  },
})
```
````

- **`gfm`** (`true` by default) to enable or disable GitHub-Flavored Markdown in MDX.
- **`smartypants`** (`true` by default) to enable or disable SmartyPants in MDX.

##### What should I do?

Delete `extendPlugins: 'markdown'` in your configuration. This is now default behavior. 

```js del={8}
// astro.config.mjs
import { defineConfig } from 'astro/config';
import mdx from '@astrojs/mdx';

export default defineConfig({
  integrations: [
    mdx({
      extendPlugins: 'markdown',
    }),
  ],
});
```

Replace `extendPlugins: 'defaults'` with `extendMarkdownConfig: false` and add the [separate options for GitHub-Flavored Markdown and SmartyPants](/en/guides/integrations-guide/mdx/#options-inherited-from-markdown-config) to enable these default plugins individually in MDX.

```js del={8} ins={9-11}
// astro.config.mjs
import { defineConfig } from 'astro/config';
import mdx from '@astrojs/mdx';

export default defineConfig({
  integrations: [
    mdx({
      extendPlugins: 'defaults',
      extendMarkdownConfig: false,
      smartypants: true,
      gfm: true,
    }),
  ],
});
```

#### Added: More MDX config options to match Markdown

> Setting `extendMarkdownConfig: false` now allows you to individually set [every available Markdown configuration option](/en/reference/configuration-reference/#markdown-options) (except `drafts`) separately in your MDX integration configuration.

Author: ```suggestion
Astro 2.0 allows you to individually set [every available Markdown configuration option](/en/reference/configuration-reference/#markdown-options) (except `drafts`) separately in your MDX integration configuration.
```

Author: Simplified this by just refering to the separate configs.

> 
> ```js
> // astro.config.mjs
> import { defineConfig } from 'astro/config';
> import mdx from '@astrojs/mdx';
> 
> export default defineConfig({
>   markdown: {
>     syntaxHighlight: 'prism',
>     remarkPlugins: [remarkPlugin1],
>     gfm: true,
>   },
>   integrations: [
>     mdx({
>       // `syntaxHighlight` inherited from Markdown
> 
>       // Markdown `remarkPlugins` ignored,
>       // only `remarkPlugin2` applied.
>       remarkPlugins: [remarkPlugin2],
>       // `gfm` overridden to `false`
>       gfm: false,
>     })
>   ]
> });
> ```
> 

Author: ``````suggestion

```js
// astro.config.mjs
import { defineConfig } from 'astro/config';
import mdx from '@astrojs/mdx';

export default defineConfig({
  markdown: {
    remarkPlugins: [remarkPlugin1],
    gfm: true,
  },
  integrations: [
    mdx({
      remarkPlugins: [remarkPlugin2],
      gfm: false,
    })
  ]
});
```

``````

#### What should I do?

 Check that both your development environment and your deployment environment are using **Node.js `16.12.0` or later**.

1. Check your local version of node using:

    ```sh
    node -v
 
    ```
    If your local development environment needs upgrading, [install Node.js](https://nodejs.org/en/download/).

2. Check your [deployment environment's](/en/guides/deploy/) own documentation to verify that they support Node 16. 

    You can specify Node `16.12.0` for your Astro project either in a dashboard configuration setting, or a `.nvmrc` file.

### Reserved: `src/content/`

Astro 2.0 now includes the Collections API for organizing your Markdown and MDX files into [content collections](/en/guides/content-collections/). This API reserves `src/content/` as a special folder.

#### What should I do?

Rename an existing `src/content/` folder to avoid conflicts. This folder, if it exists, can now only be used for [content collections](/en/guides/content-collections/).

### Changed: `Astro.site` trailing slash

In 1.0, Astro ensured the URL you set as `site` in `astro.config.mjs` always had a trailing slash when accessed using `Astro.site`.

> Astro 2.0 stops modifying the value of `site`. If you pass a URL without a trailing slash, `Astro.site` will also not have a trailing slash.

Author: ```suggestion
Astro 2.0 no longer modifies the value of `site`. `Astro.site` will use the exact value defined, and a trailing slash must be specified if desired.
```

#### What should I do?

> Make sure you set `site` using a URL that includes the trailing slash.

Author: ```suggestion
 In `astro.config.mjs`, update the URL set in `site` to include a trailing slash to match your code's v1.x behavior.
```

---
layout: ~/layouts/MainLayout.astro

> title: Upgrade Guide

Reviewer: How do you feel about changing the title here?

- more specific
- should look nice in OpenGraph images as people share this out
- I’d bet this is moderately better for SEO

```suggestion
title: Upgrade to Astro v2
```

Reviewer: And while we’re here, maybe also update the title & description of the old guide now at `upgrade-to/v1`? Less clear cut what the rename should be. Maybe something like this?

```diff
-title: Migration Guide
-description: How to migrate your project to latest version of Astro.
+title: Legacy v0.x Migration Guide 
+description: Archived guide documenting changes between pre-v1 versions of Astro
```

That page also starts with

> This guide will help you migrate from older versions of Astro to the latest version.

Probably worth updating that too.

  pnpm install astro@latest

  # Example: upgrade React and Tailwind integrations
  pnpm install @astrojs/react@latest @astrojs/tailwind@latest
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  # Upgrade to Astro v2.x
  yarn add astro@latest
  
  # Example: upgrade React and Tailwind integrations
  yarn add @astrojs/react@latest @astrojs/tailwind@latest
  ```
  </Fragment>
</PackageManagerTabs>

## Astro 2.0 Breaking Changes

Astro v2.0 includes some breaking changes, as well as the removal of some previously deprecated features. If your project doesn't work as expected after upgrading to v2.0, check this guide for an overview of all breaking changes and instructions on how to update your codebase.

See [the changelog](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md) for full release notes.

### Removed: Support for Node 14

Node 14 is scheduled to reach its End of Life in April 2023. 

Astro 2.0 drops Node 14 support entirely, so that all Astro users can take advantage of Node's more modern features.

#### What should I do?

 Check that both your development environment and your deployment environment are using **Node.js `16.12.0` or later**.

> 1. Check your local version of node using:

Reviewer 2: Generally when we write "node" in lowercase we are referring to the terminal program, so maybe we should mark it as code?

```suggestion
1. Check your local version of `node` using:
```

or just uppercase it:

```suggestion
1. Check your local version of Node using:
```

### Reserved: `src/content/`

Astro 2.0 now includes the Collections API for organizing your Markdown and MDX files into [content collections](/en/guides/content-collections/). This API reserves `src/content/` as a special folder.

#### What should I do?

Rename an existing `src/content/` folder to avoid conflicts. This folder, if it exists, can now only be used for [content collections](/en/guides/content-collections/).

### Changed: `Astro.site` trailing slash

In 1.x, Astro ensured the URL you set as `site` in `astro.config.mjs` always had a trailing slash when accessed using `Astro.site`.

Astro 2.0 no longer modifies the value of `site`. `Astro.site` will use the exact value defined, and a trailing slash must be specified if desired.

#### What should I do?

 In `astro.config.mjs`, add a trailing slash to the URL set in `site`. 

```js del={5} ins={6}
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  site: 'https://example.com',
  site: 'https://example.com/',
});
```

### Changed: `_astro/` folder for build assets

> In 1.x, assets were built to various locations, including `assets/`, `chunks/` and the root of the build output.

Reviewer 2: The sentence structure felt weird to me in the last bit, I'm wondering if there's a way to make it more seamless.

```suggestion
In 1.x, assets were built to various locations, including `assets/`, `chunks/`, and to the root of the build output.
```

Probably a "just me" thing being non-native, feel free to ignore this one if you feel it's actually ok.

Update your deployment platform configuration if it relies on the location of these assets.

### Changed: Markdown plugin configuration

#### Removed: `extendDefaultPlugins`

In 1.x, Astro used `markdown.extendDefaultPlugins` to re-enable Astro's default plugins when adding your own Markdown plugins.

Astro 2.0 removes this configuration option entirely because its behavior is now the default.

Applying remark and rehype plugins in your Markdown configuration  **no longer disables Astro's default plugins**. GitHub-Flavored Markdown and Smartypants are now applied whether or not custom `remarkPlugins` or `rehypePlugins` are configured.

##### What should I do?

Remove `extendDefaultPlugins` in your configuration. This is now Astro's default behavior in 2.0, and you can delete this line without any replacement.

```js del={6}
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  markdown: {
    extendDefaultPlugins,
  }
});
```

#### Added: `gfm` and `smartypants`

In 1.x, you could choose to disable both of Astro's default Markdown plugins (GitHub-Flavored Markdown and SmartyPants) by setting `markdown.extendDefaultPlugins: false`.

> Astro 2.0 replaces `markdown.extendDefaultPlugins: false` with separate Boolean options to control each of Astro's built-in default Markdown plugins individually. These are enabled by default, and can be set to `false` independently.

Reviewer 2: I generally prefer to put the adverb mid-sentence instead of at the end, as it makes it a bit easier to read IMO. Also, small comma change.

```suggestion
Astro 2.0 replaces `markdown.extendDefaultPlugins: false` with separate Boolean options to individually control each of Astro's built-in default Markdown plugins. These are enabled by default and can be set to `false` independently.
```

- `markdown.gfm: false` disables GitHub-Flavored Markdown
- `markdown.smartypants: false` disables SmartyPants

```js del={6} ins={7-8}
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  markdown: {
    extendDefaultPlugins: false,
    smartypants: false,
    gfm: false,
  }
});
```

### Changed: MDX plugin configuration

#### Replaced: `extendPlugins` changed to `extendMarkdownConfig`

In 1.x, the MDX integration’s `extendPlugins` option managed how your MDX files should inherit your Markdown configuration: all your Markdown configuration (`markdown`), or Astro's default plugins only (`default`).

Astro 2.0 replaces the behavior controlled by `mdx.extendPlugins` with three new, independently-configurable options that are `true` by default:

- **[`mdx.extendMarkdownConfig`](/en/guides/integrations-guide/mdx/#extendmarkdownconfig)** to inherit all or none of your Markdown configuration
- **`mdx.gfm`** to enable or disable GitHub-Flavored Markdown in MDX
- **`mdx.smartypants`** to enable or disable SmartyPants in MDX

##### What should I do?

> Delete `extendPlugins: 'markdown'` in your configuration. This is now default behavior. 

Reviewer 2: Missing "the" here?
```suggestion
Delete `extendPlugins: 'markdown'` in your configuration. This is now the default behavior.
```

In 1.0, Astro moved the old Astro-Flavored Markdown (also known as Components in Markdown) to a legacy feature.

Astro 2.0 removes the `legacy.astroFlavoredMarkdown` option completely. Importing and using components in `.md` files will no longer work.

#### What should I do?

Remove this legacy flag. It is no longer available in Astro.

```js del={3-5}
// astro.config.mjs
export default defineConfig({
	legacy: {
		astroFlavoredMarkdown: true,
	},
})

```

If you were using this feature in 1.x, we recommend [using the MDX integration](/en/guides/integrations-guide/mdx/) which allows you to combine components and JSX expressions with Markdown syntax.

### Removed: `Astro.resolve()`

In 0.24, Astro deprecated `Astro.resolve()` for getting resolved URLs to assets that you might want to reference in the browser.

Astro 2.0 removes this option entirely. `Astro.resolve()` in your code will cause an error.

#### What should I do?

Resolve asset paths using `import` instead. For example:

```astro
// src/pages/index.astro
---
import 'style.css';
import imageUrl from './image.png';
---

<img src={imageUrl} />
```

### Removed: `Astro.fetchContent()`

> In 0.26, Astro deprecated `Astro.fetchContent()`for fetching data from your local Markdown files.

Reviewer 2: Missing whitespace!
```suggestion
In 0.26, Astro deprecated `Astro.fetchContent()` for fetching data from your local Markdown files.
```

      "astro": "1.9.2",
      "react": "^18.2.0",
      "react-dom": "^18.0.0",
      "svelte": "^3.47.0"
    },
    "dependencies": {
      "@astrojs/image": "0.13.0-beta.0",
      "@astrojs/rss": "^2.0.0",
      "@astrojs/vue": "^1.2.1",
      "vue": "^3.2.41"
    }
    ```

2. Update your project's version of Astro to the latest version using your package manager. If you're using Astro integrations, please also update those to the latest version.

    <PackageManagerTabs>
      <Fragment slot="npm">
      ```shell
      # upgrades to Astro v2.x:
      npm install astro@latest
      ```
      </Fragment>
      <Fragment slot="pnpm">
      ```shell
      # upgrades to Astro v2.x:
      pnpm install astro@latest
      ```
      </Fragment>
      <Fragment slot="yarn">
      ```shell
      # upgrades to Astro v2.x:
      yarn add astro@latest
      ```
      </Fragment>
    </PackageManagerTabs>

3. Check that you are running **at least Node `16.12.0`**. In v2.0, Astro dropped support for Node `14`.

    ```sh
    node -v
    ```
    
    If necessary, [update your version of Node](#removed-support-for-node-14) in your local and deployment environments.

> ## Update your code

Reviewer 3: this is also confusing: how are these different from the next section of "Breaking Changes"? Are these different / duplicate, or do I need to read both?

Many of these seem like duplicates of the items below. If that's the case, then this section is safe to remove and then rewrite the next section to add:

```diff

## Astro 2.0 Breaking Changes

+ Astro v2.0 includes some breaking changes, as well as the removal of some previously deprecated features. If your project doesn't work as expected after upgrading to v2.0, check this guide for an overview of all breaking changes and instructions on how to update your codebase.

+ See [the changelog](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md) for full release notes.

### Updated: Vite 4
```

Remember: A migration guide's goal isn't to summarize/highlights every single change that we've made. It's goal is to describe what a user needs to do to migrate their project from v1 -> v2, as concisely as possible. I get a vibe of "selling v2" from this section, but that's not necessary and gets in the way by making the upgrade seem more breaking/confusing than it needs.
